src/v/raft/tests/stm_manager_test.cc

```diff
@@ -0,0 +1,392 @@
+#include "raft/types.h"
+#include "random/generators.h"
+#include "serde/serde.h"
+#include "storage/record_batch_builder.h"
+#include "storage/types.h"
+#include "test_utils/async.h"
+#include "test_utils/randoms.h"
+#include "test_utils/test.h"
+
+#include <seastar/core/circular_buffer.hh>
+#include <seastar/core/io_priority_class.hh>
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/coroutine/parallel_for_each.hh>
+#include <seastar/util/log.hh>
+
+using namespace raft;
+namespace {
+static ss::logger logger("stm-test-logger");
+}
+
+/**
+ * Simple kv-store state machine
+ */
+struct simple_kv : public raft::state_machine_base {
+    using state_t = absl::flat_hash_map<ss::sstring, ss::sstring>;
+    explicit simple_kv(raft_node_instance& rn)
+      : raft_node(rn) {}
+
+    static void apply_to_state(
+      const model::record_batch& batch,
+      absl::flat_hash_map<ss::sstring, ss::sstring>& state) {
+        if (batch.header().type != model::record_batch_type::raft_data) {
+            return;
+        }
+        batch.for_each_record([&state](model::record r) {
+            auto k = serde::from_iobuf<ss::sstring>(r.key().copy());
+            auto v = serde::from_iobuf<std::optional<ss::sstring>>(
+              r.value().copy());
+            if (v) {
+                state[k] = *v;
+            } else {
+                state.erase(k);
+            }
+        });
+    }
+    ss::future<> start() override { return ss::now(); }
+    ss::future<> stop() override { co_await raft::state_machine_base::stop(); };
+
+    ss::future<> apply(const model::record_batch& batch) override {
+        apply_to_state(batch, state);
+        co_return;
+    }
+
+    ss::future<> apply_raft_snapshot(const iobuf& buffer) override {
+        state = serde::from_iobuf<state_t>(buffer.copy());
+        co_return;
+    };
+
+    const char* get_name() const override { return "simple_kv"; };
+
+    ss::future<iobuf>
+    take_snapshot(model::offset last_included_offset) override {
+        state_t inc_state;
+        // build incremental snapshot
+        auto snap = co_await raft_node.raft()->open_snapshot();
+        auto start_offset = raft_node.raft()->start_offset();
+        if (snap) {
+            auto data = co_await read_iobuf_exactly(
+              snap->reader.input(), co_await snap->reader.get_snapshot_size());
+            inc_state = serde::from_iobuf<state_t>(std::move(data));
+        }
+
+        auto rdr = co_await raft_node.raft()->make_reader(
+          storage::log_reader_config(
+            start_offset, last_included_offset, ss::default_priority_class()));
+
+        auto batches = co_await model::consume_reader_to_memory(
+          std::move(rdr), model::no_timeout);
+
+        std::for_each(
+          batches.begin(), batches.end(), [this](const model::record_batch& b) {
+              return simple_kv::apply_to_state(b, state);
+          });
+
+        co_return serde::to_iobuf(std::move(state));
```

Comment:
Should this be applying state to `inc_state` and returning `inc_state`?

Can we make a test that catches this?

Reply:
i've change the test and made kv-store operations not idempotent this way the test now catches this issue.